test/e2e/proxy_test.go

```diff
@@ -1084,3 +1078,153 @@ func testProxyServiceLifeCycle(ipFamily *corev1.IPFamily, ingressIPs []string, d
+
+			// Create another netns to fake an external network on the host network Pod.
+			externalClient := randName("external-client-")
+			cmd, externalNetns := getCommandInFakeExternalNetwork("sleep infinity", externalIPPrefix, externalClientIP, externalClientGateway)
+			if err := NewPodBuilder(externalClient, data.testNamespace, toolboxImage).OnNode(ingressNode).WithCommand([]string{"sh", "-c", cmd}).InHostNetwork().Privileged().Create(data); err != nil {
+				t.Fatalf("Failed to create external client: %v", err)
+			}
+			defer deletePodWrapper(t, data, data.testNamespace, externalClient)
+			if err := data.podWaitForRunning(defaultTimeout, externalClient, data.testNamespace); err != nil {
+				t.Fatalf("Error when waiting for Pod '%s' to be in the Running state", externalClient)
+			}
+
+			// Since the "external client" runs in a netns on ingress Node, we install a route on backend Node to route reply
+			// traffic to ingress Node.
+			addRouteToClientIPCmd := []string{"ip", "route", "replace", externalClientIP, "via", ingressNodeIP}
+			stdout, stderr, err := data.RunCommandFromAntreaPodOnNode(backendNode1, addRouteToClientIPCmd)
+			require.NoError(t, err, "Failed to add route to client IP on Node %s, stdout: %s, stderr: %s", backendNode1, stdout, stderr)
+			stdout, stderr, err = data.RunCommandFromAntreaPodOnNode(backendNode2, addRouteToClientIPCmd)
+			require.NoError(t, err, "Failed to add route to client IP on Node %s, stdout: %s, stderr: %s", backendNode2, stdout, stderr)
+			defer func() {
+				delRouteToClientIPCmd := []string{"ip", "route", "del", externalClientIP, "via", ingressNodeIP}
+				stdout, stderr, err := data.RunCommandFromAntreaPodOnNode(backendNode1, delRouteToClientIPCmd)
+				require.NoError(t, err, "Failed to delete route to client IP on Node %s, stdout: %s, stderr: %s", backendNode1, stdout, stderr)
+				stdout, stderr, err = data.RunCommandFromAntreaPodOnNode(backendNode2, delRouteToClientIPCmd)
+				require.NoError(t, err, "Failed to delete route to client IP on Node %s, stdout: %s, stderr: %s", backendNode2, stdout, stderr)
+			}()
+
+			serviceName := fmt.Sprintf("svc-dsr")
+			annotations := map[string]string{
+				types.ServiceLoadBalancerModeAnnotationKey: "dsr",
+			}
+			service, err := data.createAgnhostLoadBalancerService(serviceName, tc.withSessionAffinity, false, []string{lbIP}, &ipProtocol, annotations)
+			require.NoError(t, err)
+			defer data.deleteServiceAndWait(defaultTimeout, serviceName, data.testNamespace)
+
+			curlServiceWithPath := func(clientPod, clientNetns, path string) string {
+				testURL := fmt.Sprintf("http://%s:%d/%s", lbIP, service.Spec.Ports[0].Port, path)
+				cmd = fmt.Sprintf("curl --connect-timeout 1 --retry 5 --retry-connrefused %s", testURL)
+				if clientNetns != "" {
+					cmd = fmt.Sprintf("ip netns exec %s %s", clientNetns, cmd)
+				}
+				stdout, stderr, err = data.RunCommandFromPod(data.testNamespace, clientPod, "toolbox", []string{"sh", "-c", cmd})
```

Comment:
Is the `sh` shell actually needed here?

Reply:
It's needed. Otherwise the whole command would be regarded as an executable and kubelet would try to find it:

Internal error occurred: error executing command in container: failed to exec in container: failed to start exec "3ef22b5120aaef0fe232e01430ef127e2035e4a64a3e51d9705331c29338eeee": OCI runtime exec failed: exec failed: unable to start container process: exec: "ip netns exec ext-ehw34 curl --connect-timeout 1 --retry 5 --retry-connrefused http://1.1.2.1:8080/clientip": stat ip netns exec ext-ehw34 curl --connect-timeout 1 --retry 5 --retry-connrefused http://1.1.2.1:8080/clientip: no such file or directory: unknown

Reply:
That's surprising to me. Even when the command is split and passed as a slice? `[]string{"ip", "netns", ...}`

Reply:
Splitting the command into a slice could work. Both of them are commonly used to set command, `[]string{"sh", "-c", cmd}` is easier to construct complex command and the formatting is more friendly in some cases.